playbot/playbot.py

```diff
@@ -59,48 +60,55 @@ def name(self):
     def login(self):
         if not self.game:
             return
-        sleep(0.2)
+        sleep(0.1)
         self.send('takeover')
-        sleep(0.2)
+        sleep(0.1)
         data = self.get_data()
-        if b'Name?' in data:
+        if 'Name?' in data:
             self.name()
         else:
             self.game = False
 
     def heartbeat_socket(self):
-        if self.heartbeat > 1500:
+        if self.heartbeat > 15000:
             self.send('ack')
             self.heartbeat = 0
         else:
             self.heartbeat += 1
 
     def get_data(self):
+        buffor = ''
         try:
-            self.heartbeat_socket()
-            data = self.socket.recv(1024)
-            return data
-        except socket.timeout:
-            return b''
+            while True:
+                self.heartbeat_socket()
+                buffor += self.socket.recv(1).decode('utf-8')
+                if '\x04' in buffor:
+                    self.heartbeat = 0
+                    return buffor.split('\x04')[0]
+        except BlockingIOError:
+            self.game = True
         except ConnectionResetError as e:
             self.log(e.strerror)
             self.game = False
-            return b''
         except ConnectionAbortedError as e:
             self.log(e.strerror)
             self.game = False
```

Comment:
Clear buffor when you catch execption so there will not be risk that you will return part of message (for example exception will occure in middle of data read). In play() you check if data startwith which can be true even after you didn't read the whole message.

Reply:
make sense - will be improved.